### PR TITLE
Adds Serializer and virtual functions for new transform inits

### PR DESCRIPTION
**Pull request by SteveBronder**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

This adds a `stan::io::serializer` class that reads in objects and writes them to a linear array of data. This will be used in the new `transform_inits()` signature for the model class to serialize the unconstrained parameters.

The only function here is `.write()` which accepts any Stan type and writes that to the internal reference to a linear vector.

```cpp
Eigen::MatrixXd xx = Eigen::MatrixXd::Random(10, 10)
Eigen::VectorXd my_vec(100);
stan::io::serializer serializer(my_vec);
serializer.write(xx); // Write xx to my_vec
```

The new transform inits signature is of the form

```
void transform_inits(Eigen::VectorXd& input_r, 
                                Eigen::VectorXd& params_r,
                                std::ostream* msgs)
```

Where `input_r` is the input parameters values and  `params_r` is an input/output vector of the unconstrained parameters values

The new transform inits will use these classes by wrapping `input_r` into `stan::io::deserializer` and `params_r` into `stan::io::serializer`, where it can perform the unconstrain as

```cpp
stan::io::deserializer deserializer(input_r);
stan::io::serializer serializer(params_r);
serializer.write(deserializer.read_free_lub<std::vector<Eigen::MatrixXd>>(lb, ub, J, N, M));
```

Where `J` is the size of the array and `N`, `M` are the dimensions of the matrix.

#### Intended Effect

Along with #3018 this will cleanup a lot of the code generated by the Stan compiler

#### How to Verify

Tests can be run with

```bash
./runTests.py ./src/test/unit/io/serializer_varmat_test.cpp ./runTests.py ./src/test/unit/io/serializer_test.cpp ./runTests.py ./src/test/unit/io/serializer_stdvector_test.cpp
```

These test all the basic types, arrays of those basic types, and `var_value<>` types with inner values of the basic types

#### Side Effects

None

#### Documentation

Documentation is written for the new class and for each method in the class.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Steve Bronder

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses: Ste
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Review by bbbales2 — CHANGES_REQUESTED** [review]
I like the serializer and updated the tests a bit. That part is good.

I left a Q about one of the transform inits.

It seems we're getting warnings in upstream tests about this too that are causing Jenkins to reject this:

```
'transform_inits' overrides a member function but is not marked 'override' [-Winconsistent-missing-override] void transform_inits(const io::var_context& context, ^ src/test/test-models/transformed_data_rng_test.hpp:37:54: note: in instantiation of template class 'stan::model::model_base_crtp' requested here class transformed_data_rng_test_model final : public model_base_crtp { ^
```

File: src/stan/model/model_base_crtp.hpp [hunk 1]
```diff
@@ -199,6 +199,18 @@ class model_base_crtp : public stan::model::model_base {
     return static_cast<const M*>(this)->template write_array(
         rng, theta, theta_i, vars, include_tparams, include_gqs, msgs);
   }
+
+  void transform_inits(const io::var_context& context,
+                       Eigen::VectorXd& params_r, std::ostream* msgs) const {
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply src/stan/model/model_base_crtp.hpp]: Oddly I'm not seeing it in `model_base_crtp` but am seeing it in `model_base`. I think it just needs an override keyword and we'll miss this

**Comment by SteveBronder** [comment]
@bbbales2 I thought we could just add the new signature in this PR but I think the steps here are

1. Put in this PR and the deserializer PR
2. Ryan puts in his PR
3. We then add the new `transform_init()` signature since it exists in the model and use it in the code

**Comment by SteveBronder** [comment]
Aight @bbbales2 I think this is good to go!

**Review by bbbales2 — APPROVED** [review]
Good!

**Comment by seantalts** [comment]
Are there examples of what the new stanc3 code should look like using this? 

Just a note from another thread - we can't change the `transform_init` method without bumping the major version number of Stan, so I would propose we not plan on that right now.

**Pull request merged**